addons/udes_security/models/res_archiving_restrictions.py

```diff
@@ -0,0 +1,82 @@
+    Configurable access to _only_ record archiving & unarchiving
+
+    If an Archiving Restriction is created for a model, the group_ids set will determine
+    which groups have the facility to archive and unarchive records on the model.
+
+    This includes backend writes to the `active` field, as well as hiding
+    the Archive/Unarchive action from the model, and attempting to remove active
+    fields & smart buttons from form views.
+    """
+
+    _name = "res.archiving.restrictions"
+    _description = "Record Archiving Restrictions"
+    _sql_constraints = [
+        (
+            "model_id_uniq",
+            "unique(model_id)",
+            _(
+                "An Archiving Restriction already exists for this model. Please edit the existing one"
+            ),
+        )
+    ]
+
+    model_id = fields.Many2one(
+        "ir.model", string="Model to apply restriction to", required=True, index=True
+    )
+    name = fields.Char(string="Model Name", related="model_id.model", readonly=True)
+    group_ids = fields.Many2many(
+        "res.groups",
+        string="Groups permitted to archive & unarchive",
+        index=True,
+        help="""
+        If no groups are configured, ALL users will be prevented from archiving (except superuser)
+        If the user is in any of the configured groups, they will be permitted access
+        If no restriction on a model is intended, then do not create a restriction record for that model
+        """,
+    )
+
+    def _get_can_archive(self, model_name):
+        """
+        Test against the model to see if our user is
+        permitted to archive/unarchive the record from the server side.
+
+        Can archive if user has _any_ groups defined on the restriction,
+        or no restriction is defined
+
+        :param: model_name: str() model _name i.e "stock.picking.type"
+        :returns: bool()
+        """
+        ResArchivingRestrictions = self.env["res.archiving.restrictions"]
+
+        if self.env.user.id == SUPERUSER_ID:
+            return True
+        if not model_name:
+            return True
+        model_restriction = ResArchivingRestrictions.search([("model_id.model", "=", model_name)])
+        if not model_restriction:
+            return True
+        user_group_ids = self.env.user.groups_id.ids
+        groups_on_user_and_restriction = [
+            gid for gid in user_group_ids if gid in model_restriction.group_ids.ids
+        ]
+        return bool(groups_on_user_and_restriction)
+
+    def _check_can_archive(self, model_name):
+        """Raises a warning if the logged in user does not meet
+        the group requirements to archive/unarchive a record
+
+        :param: model_name: str() model _name i.e "stock.picking.type"
+        """
+        if not self._get_can_archive(model_name):
+            raise ValidationError(
+                _(
+                    f"Cannot archive/unarchive record due user not meeting restrictions in place on {model_name}"
```

Comment:
Suggestion: As this will be a user facing error if it ever comes up (which it shouldn't) I'd suggest rewording it slightly, e.g.:

`You do not have permission to archive/unarchive this record.`